pyccl/base/deprecations.py

```diff
@@ -73,35 +79,40 @@ def wrapper(*args, **kwargs):
-        # if ("pyccl.halos" in func.__module__
-        #         and func.__name__ == "__init__"
-        #         and is_instance(catch_cosmo, "Cosmology")):
-        #     warnings.warn(
-        #         f"Use of argument `cosmo` has been deprecated in {name}. "
-        #         "This will trigger an exception in the future.",
-        #         CCLDeprecationWarning)
-        #     # `cosmo` may be in `args` or in `kwargs`, so we check both.
-        #     args = tuple(
-        #         item for item in args if not is_instance(item, "Cosmology"))
-        #     kwargs.pop("cosmo", None)
-        #
+        catch_cosmo = args[1] if len(args) > 1 else kwargs.get("cosmo")
+        if ("pyccl.halos" in func.__module__
+                and func.__name__ == "__init__"
+                and is_instance(catch_cosmo, "Cosmology")):
+            warnings.warn(
+                f"Use of argument `cosmo` has been deprecated in {name}. "
+                "This will trigger an exception in the future.",
+                CCLDeprecationWarning)
+            # `cosmo` may be in `args` or in `kwargs`, so we check both.
+            args = tuple(
+                item for item in args if not is_instance(item, "Cosmology"))
+            kwargs.pop("cosmo", None)
+
         # API compatibility for reordered positionals in `fourier_2pt`.
-        # first_arg = args[1] if len(args) > 1 else None
-        # if (func.__name__ == "fourier_2pt"
-        #         and is_instance(first_arg, "HaloProfile")):
-        #     api = dict(zip(["prof", "cosmo", "k", "M", "a"], args[1: 6]))
-        #     args = (args[0],) + args[6:]  # discard args [1-5]
-        #     kwargs.update(api)            # they are now kwargs
-        #     warnings.warn(
-        #         "API for Profile2pt.fourier_2pt has changed. "
-        #         "Argument order (prof, cosmo, k, M, a) has been replaced by "
-        #         "(cosmo, k, M, a, prof).", CCLDeprecationWarning)
+        first_arg = args[1] if len(args) > 1 else None
+        if (func.__name__ == "fourier_2pt"
+                and is_instance(first_arg, "HaloProfile")):
+            api = dict(zip(["prof", "cosmo", "k", "M", "a"], args[1: 6]))
+            args = (args[0],) + args[6:]  # discard args [1-5]
+            kwargs.update(api)            # they are now kwargs
+            warnings.warn(
+                "API for Profile2pt.fourier_2pt has changed. "
+                "Argument order (prof, cosmo, k, M, a) has been replaced by "
+                "(cosmo, k, M, a, prof).", CCLDeprecationWarning)
 
         # API compatibility for renamed arguments.
         warn_names = set(kwargs) - set(params)
+        unexpected = [k for k in warn_names if k not in rename]
+        if unexpected:
+            # emulate Python default behavior for arguments that don't exist
+            raise TypeError(
+                f"{func.__name__}() got an unexpected keyword argument "
+                f"'{unexpected[0]}'")
```

Comment:
If you had this function:
```python
def func(a, b, c):
    ...
```
and you wanted to rename your arguments like this:
```python
def func(a, bee, see):
    pass
```
These calls would be valid:
```python
func(a=1, bee=2, see=3)  # new API
func(a=1, b=2, c=3)  # old API - throws a warning
```
But this call would throw a weird `KeyError` because it would internally try to find the renamed argument but wouldn't be able to find it:
```python
func(a=1, b=2, c=3, hello="hi")
```
So these lines emulate what Python would say if you put something that shouldn't be there.